Feature/download component

bkheadwire opened the pull request:
<!--
Before making a PR please make sure to read our contributing guidelines
https://github.com/adobe/aem-core-wcm-components/blob/master/CONTRIBUTING.md

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/)
followed by the ticket number fixed by the PR. It should be underlined in the preview if done correctly.
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | [[RTC] Download Component](https://github.com/adobe/aem-core-wcm-components/issues/217)
| Patch: Bug Fix?          | N/A?
| Minor: New Feature?      | Yes
| Major: Breaking Change?  | No
| Tests Added + Pass?      | Only minimally enough to get it to build
| Documentation Provided   | No readme yet but some comments (mostly javadocs for the model)
| Any Dependency Changes?  | No
| License                  | Apache License, Version 2.0

<!-- Describe your changes below in as much detail as possible -->

Was asked to make a PR for this implementation of the download component so the core team can take a look at it. It's definitely still a WIP.

kwin commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java:
```diff
@@ -0,0 +1,179 @@
+import com.day.cq.commons.DownloadResource;
+import com.day.cq.dam.api.Asset;
+import com.day.cq.wcm.api.designer.Style;
+import com.day.crx.JcrConstants;
```
Wrong package? Rather use constant from Javax.jcr.Property or from https://helpx.adobe.com/experience-manager/6-4/sites/developing/using/reference-materials/javadoc/com/day/cq/commons/jcr/JcrConstants.html

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java:
```diff
@@ -0,0 +1,179 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(DownloadImpl.class);
+
+    public final static String RESOURCE_TYPE = "mnrdlm/brand/components/content/download";
```
Should be: `core/wcm/components/download/v1/download`

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java:
```diff
@@ -0,0 +1,179 @@
+    @Nonnull
+    @Override
+    public String getExportedType() {
+        return request.getResource().getResourceType();
```
I'd use `resource.getResourceType()`.

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java:
```diff
@@ -0,0 +1,102 @@
+ */
+public interface Download extends ComponentExporter {
+
+    String PN_TITLE_FROM_ASSET = "titleFromAsset";
```
Same as above, I'd call it `useTitleFromAsset`.

richardhand replied:
@jckautzmann - I think it's clear enough, and it follows our naming in the teaser [0]. What we are missing here are some docs for these properties as we have in the other component interfaces.

[0] https://github.com/adobe/aem-core-wcm-components/blob/master/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Teaser.java#L109

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java:
```diff
@@ -0,0 +1,102 @@
+public interface Download extends ComponentExporter {
+
+    String PN_TITLE_FROM_ASSET = "titleFromAsset";
+    String PN_DESCRIPTION_FROM_ASSET = "descriptionFromAsset";
```
Same as above, I'd call it `useDescriptionFromAsset`.

jckautzmann replied:
Please add javadocs.

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java:
```diff
@@ -0,0 +1,102 @@
+ */
+public interface Download extends ComponentExporter {
+
+    String PN_TITLE_FROM_ASSET = "titleFromAsset";
```
Please add javadocs.

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java:
```diff
@@ -0,0 +1,179 @@
+        descriptionFromAsset = properties.get(PN_DESCRIPTION_FROM_ASSET, descriptionFromAsset);
+        if (currentStyle != null) {
+            if (StringUtils.isBlank(ctaText)) {
+                ctaText = currentStyle.get("ctaText", String.class);
```
we should define an API constant for `ctaText`.

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java:
```diff
@@ -0,0 +1,179 @@
+            if (StringUtils.isBlank(ctaText)) {
+                ctaText = currentStyle.get("ctaText", String.class);
+            }
+            titleType = currentStyle.get("titleType", String.class);
```
we should define an API constant for `titleType`.

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java:
```diff
@@ -0,0 +1,179 @@
+
+                    String resourcePath = resourceResolver.map(request, resource.getPath());
+
+                    imagePathBuilder.append(resourcePath).append(".coreimg.jpeg");
```
we should define a constant for the string ".coreimg.jpeg".

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java:
```diff
@@ -0,0 +1,179 @@
+
+                    imagePathBuilder.append(resourcePath).append(".coreimg.jpeg");
+                    if (lastModified > 0) {
+                        imagePathBuilder.append("/").append(lastModified).append(".jpeg");
```
we should define a constant for the string ".jpeg".

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java:
```diff
@@ -0,0 +1,179 @@
+                    imagePath = imagePathBuilder.toString();
+
+                    if (titleFromAsset) {
+                        String assetTitle = downloadAsset.getMetadataValue("dc:title");
```
we should define a constant for the string "dc:title".

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java:
```diff
@@ -0,0 +1,179 @@
+                        }
+                    }
+                    if (descriptionFromAsset) {
+                        String assetDescription = downloadAsset.getMetadataValue("dc:description");
```
we should define a constant for the string "dc:description".

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @return the asset url
+     */
+    default String getDownloadUrl(){
```
I'd call it `getUrl()` to be consistent with other method names like `getTitle`, `getDescription`.

jckautzmann commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    /**
+     * Returns the element to be used for the title as defined in the component policy.
```
I would specify: `the HTML element to be used...`

jckautzmann commented on bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImplTest.java:
```diff
@@ -0,0 +1,123 @@
+        return getDownloadUnderTest(resourcePath, null);
+    }
+
+    private Download getDownloadUnderTest(String resourcePath, Style currentStyle) {
```
This block could be simplified: see for example: https://github.com/adobe/aem-core-wcm-components/blob/development/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v1/CarouselImplTest.java#L82

jckautzmann commented on content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/.content.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    jcr:description="Creates a tile for a downloadable dam asset"
```
DAM should be upper case.

jckautzmann commented on content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/_cq_dialog/.content.xml:
```diff
@@ -0,0 +1,255 @@
+                                                        fieldDescription="A title to display as the headline for the download."
+                                                        fieldLabel="Title"
+                                                        name="./jcr:title"
+                                                        value="${cqDesign._jcr_description}"/>
```
Shouldn't it be `cqDesign._jcr_title`?

jckautzmann commented on content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/clientlibs/editor/js/download.js:
```diff
@@ -0,0 +1,96 @@
+        if (dialogContent) {
+            titleTuple = new CheckboxTextfieldTuple(dialogContent, titleCheckboxSelector, titleTextfieldSelector);
+            descriptionTuple = new CheckboxTextfieldTuple(dialogContent, descriptionCheckboxSelector, descriptionTextfieldSelector, true);
+            $cqFileUpload     = $dialog.find(".cq-FileUpload");
```
I'd define constants for all the strings:
".cq-FileUpload", ".cq-FileUpload-edit", "assetselected", "[coral-fileupload-clear]", "cqFileuploadFilereference", "/_jcr_content/metadata.json", "dc:description", "dc:title"

jckautzmann commented:
@bkheadwire Thx for your PR! I went through the code and added a few comments. I installed it on a local instance: it looks good.
When playing with the component, I noticed an issue when scrolling the Title Type dropdown of the policy dialog.
I only tested with a .jpg asset from DAM. Does the component support other mime types? Videos, documents (.doc, .pdf)?
This component still needs to support defining file name, file format, file size: are you intending to add this to the PR?

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/download.html:
```diff
@@ -0,0 +1,31 @@
+    <div class="cmp-download__description" data-sly-test.description="${download.description}">
+        ${description @ context="html"}
+    </div>
+    <a class="cmp-download__cta" href="${download.downloadUrl}">${download.ctaText}</a>
```
In the teaser component we refer to CTAs as "action" in code and "Call-To-Action" in the dialog - we should be consistent here.

richardhand commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java:
```diff
@@ -0,0 +1,102 @@
+     * Returns either the description configured in the dialog or the title of the DAM asset,
+     * depending on the state of the descriptionFromAsset checkbox.
+     *
+     * @return the download title
```
"the download description"

richardhand commented on bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    /**
+     * Returns either the description configured in the dialog or the title of the DAM asset,
```
"description of the DAM asset"

Although I'm also not sure we should reveal the implementation details in these interface descriptions (dialog fields, types etc.) - I also would be inclined not to mention how the title is retrieved, which may (although unlikely) be subject to change in future version implementations. Same comment for the description.

bkheadwire commented:
> @bkheadwire Thx for your PR! I went through the code and added a few comments. I installed it on a local instance: it looks good.

Thank you for all of the detailed feedback! I've addressed many of the items you mentioned and will continue to do so as I have time.

> When playing with the component, I noticed an issue when scrolling the Title Type dropdown of the policy dialog.

Yep, I see that. It looks like the entire panel is trying to scroll in addition to the dropdown itself. Haven't had a chance to look at that yet.

> I only tested with a .jpg asset from DAM. Does the component support other mime types? Videos, documents (.doc, .pdf)?

When the asset is dropped from the DAM, every format I've tried has at least been downloadable (which isn't particularly surprising - currently the download link points straight at the asset and exposes the DAM structure which is something I've gotten some feedback from @vladbailescu about changing). However, if the asset doesn't have a usable image rendition, the thumbnail will be broken. There's also currently a problem with the dialog not properly loading the thumbnail or the properties from the DAM after it's been closed and reopened.

The browse functionality currently doesn't work at all. :-1: 

> This component still needs to support defining file name, file format, file size (according to the requirements in #217): are you intending to add this to the PR?

Yes!

kwin commented:
Currently this PR does not set a content disposition header (which would require a dedicated servlet, compare with https://github.com/adobe/aem-core-wcm-components/issues/217#issuecomment-402534042). Also in some customer setups the direct access to DAM paths/servlets is prohibited by the dispatcher.

vladbailescu commented:
Mergin this PR into the `feature/download` branch for now. Will add improvements on top of that.
